scripts/specializations/CpCourseGenerator.lua

```diff
@@ -0,0 +1,86 @@
+--- Specialization implementing the course generator related functionality
+---
+---@class CpCourseGenerator
+CpCourseGenerator = {}
+
+CpCourseGenerator.MOD_NAME = g_currentModName
+CpCourseGenerator.NAME = ".cpCourseGenerator"
+CpCourseGenerator.SPEC_NAME = CpCourseGenerator.MOD_NAME .. CpCourseGenerator.NAME
+
+function CpCourseGenerator.register(typeManager,typeName,specializations)
+    if CpCourseGenerator.prerequisitesPresent(specializations) then
+        typeManager:addSpecialization(typeName, CpCourseGenerator.SPEC_NAME)
+    end
+end
+
+function CpCourseGenerator.prerequisitesPresent(specializations)
+    return SpecializationUtil.hasSpecialization(CpAIWorker, specializations)
+end
+
+function CpCourseGenerator.registerEventListeners(vehicleType)
+    SpecializationUtil.registerEventListener(vehicleType, "onUpdate", CpCourseGenerator)
+    SpecializationUtil.registerEventListener(vehicleType, "onLoad", CpCourseGenerator)
+end
+
+function CpCourseGenerator.registerFunctions(vehicleType)
+    SpecializationUtil.registerFunction(vehicleType, 'cpDetectFieldBoundary', CpCourseGenerator.cpDetectFieldBoundary)
+    SpecializationUtil.registerFunction(vehicleType, 'cpGetFieldPolygon', CpCourseGenerator.cpGetFieldPolygon)
+    SpecializationUtil.registerFunction(vehicleType, 'cpDrawFieldPolygon', CpCourseGenerator.cpDrawFieldPolygon)
+end
+
+function CpCourseGenerator:onLoad(savegame)
+    -- create shortcut to this spec
+    self.spec_cpCourseGenerator = self["spec_" .. CpCourseGenerator.SPEC_NAME]
```

Comment:
Do we want to hold a tempory boundary, which was generated for example on changing the start position marker in the gui and also keep a fixed boundary each time we generate a course, which is not overwritten by the action above?

Reply:
Maybe we can save/load and mp sync those field boundaries later for vehicle which have a current course loaded or do want to generate boundaries again on job start, if needed for the drive strategy?

Reply:
For the sake of simplicity, I'd re-detect the boundary on every generator session. We already store a field polygon in the AIJob, and I think it is used only by the unloader, for combine self unload and the bale finder. We do not store that field polygon in the savegame, instead, we re-scan the field if we don't have the polygon anymore, which is fast, as we don't do island detection. I'd prefer here rescanning as well, but the asynchronous nature of the Giant detector would probably make it difficult, also not sure if island detection can be disabled there.

Reply:
@pvaiko You've got a good point there, as long as we don't need the perfect field bondary from the generation for the fieldwork job, than we can ignore all this and work as before :)